Jetpack Boost: Improved page cache and error handling

pyronaur opened the pull request:
Fixes #35896

## Proposed changes:
- Enhanced the page cache module state management and error messaging in Jetpack Boost
- Implemented a more responsive UI that reduces flickering
- Added actions to clear the cache and improved handling for setting up the cache
- Refactored error handling to restore proper message formatting
- Introduced a new 'dismissed' property for page cache errors to manage their visibility
- Removed the 'Meta' component for page cache and its associated styles, concentrating logic into the 'PageCache' component
- Added a callback for actions to take before toggling the page cache module
- Updated handling of cache bypass patterns and logging within the page cache module

### Other information:
- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?



## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
- Ensure that the page cache module can be enabled and disabled without issues
- Verify that error messages display correctly and can be dismissed
- Check the responsiveness of the UI, ensuring there's no excessive flickering during operations
- Test the error handling by introducing known errors and observing correct message formatting

pyronaur committed:
Restore error formatting

pyronaur committed:
Add plugins/boost changelog entry.

pyronaur committed:
Fix modulestate type

pyronaur committed:
Auto-disble page cache module on error

pyronaur committed:
Prevent excessive flickering

pyronaur committed:
Dedupe requests, more responsive UI

pyronaur committed:
Add plugins/boost changelog entry.

pyronaur committed:
Merge branch 'trunk' into fix/boost/page-cache-error-handling

# Conflicts:
#	projects/plugins/boost/app/assets/src/js/features/page-cache/health/health.tsx
#	projects/plugins/boost/app/assets/src/js/features/page-cache/health/lib/get-error-data.tsx
#	projects/plugins/boost/app/assets/src/js/features/page-cache/page-cache.tsx

pyronaur committed:
unused async

pyronaur committed:
monolintfix

dilirity commented on projects/plugins/boost/app/assets/src/js/features/page-cache/health/health.tsx:
```diff
@@ -1,34 +1,61 @@
+				<MutationNotice
+					mutationId="page-cache-setup"
+					isPending={
+						setup.isPending || ( !! pageCache?.active && ( ! error || !! error.dismissed ) )
```
I think this causes the `Setting up cache...` to show up when opening the page and caching is enabled (and working).

![image](https://github.com/Automattic/jetpack/assets/11799079/f291916d-3067-42d6-bf54-6b6da6b8886e)

To reproduce, enable caching and wait for setup to complete. Then refresh the page and you should see the same notice.

dilirity replied:
With 112d204 this no longer happens.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/page-cache/health/health.tsx:
```diff
@@ -1,34 +1,61 @@
+						setup.isPending || ( !! pageCache?.active && ( ! error || !! error.dismissed ) )
+					}
+					isError={ setup.isError }
+					isSuccess={ setup.isSuccess && ! error }
```
Is it possible to have setup success and an error? Just wondering why `! error` is needed here.

pyronaur replied:
This is to prevent sticky `isSuccess` true on `setup` when toggling on again.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/page-cache/health/lib/get-error-data.tsx:
```diff
@@ -189,13 +189,10 @@ type FormattedError = {
 };
-function getErrorData( status?: PageCacheError ): FormattedError {
+function getErrorData( status?: PageCacheError ): false | FormattedError {
 	// Fall back to unknown error if no status
```
This comment is now outdated.

dilirity commented on projects/plugins/boost/app/assets/src/js/pages/index/index.tsx:
```diff
@@ -154,10 +164,15 @@ const Index = () => {
 				}
 			>
-				<PageCache setup={ pageCacheSetup } error={ pageCacheError.data } />
+				{ ! pageCacheError.data && ! pageCacheSetup.isError && <PageCache /> }
```
Bye bye prop drilling. Nice!

dilirity reviewed (dismissed):
Nice work on this!

There's just one small thing and we can 🚢

pyronaur committed:
Remove outdated comment

pyronaur committed:
Remove extra pagecache active checks for now

dilirity committed:
Make linter happy

dilirity committed:
Show notice when enabling Page Cache module (before setup is ran)

dilirity committed:
Remove page cache setup notice before enabling page cache

dilirity commented:
@pyronaur I hope you don't mind, but I did some updates to the feedback a user would get when working with the page cache module.

Previously, if on a slow connection and you enable the page cache module, the mutation notice for setting up cache would show up after the module's `set` request completed. During that time, the meta (summary, actions) would be visible and a user might think everything is okay and navigate away.

I've added a "Enabling page cache..." mutation notice right when the user toggles the module, to show an indication that something's happening right away. As soon as the `setup` request is sent, that mutation notice is removed and the "Setting up cache...` one is shown.

The second thing I fixed is when there was a setup error, you'd get the `An error occurred while setting up cache.` mutation notice. After toggling the module, the notice under the description disappears, but the mutation notice remained showing the `An error occurred while setting up cache.` message.

I've updated this so as soon as the user clicks to enable the module, the `setup` mutation notice is removed.

Please review the changes, as I'm not sure what I did is the best approach.

thingalon commented:
We've got a little style breakage here: 
![image](https://github.com/Automattic/jetpack/assets/1369626/610a4c7a-48f2-4aa3-bfa1-370a35ab801f)

thingalon reviewed (dismissed):
Seems to work well :)

thingalon commented on projects/plugins/boost/app/assets/src/js/features/page-cache/health/health.tsx:
```diff
@@ -1,34 +1,62 @@
-	}
 
-	return null;
+	const [ canReset, setCanReset ] = useState( false );
+
+	useEffect( () => {
+		if ( setup?.isError && error && ! error.dismissed ) {
+			setCanReset( true );
+		}
+	}, [ setup?.isError, error, setCanReset ] );
+
+	useEffect( () => {
+		if ( canReset ) {
+			setCanReset( false );
+			setPageCache( false );
+		}
+	}, [ canReset, setCanReset, setPageCache ] );
```
I think a coment explaining canReset would be a good idea. Can be done in a followup PR.

dilirity committed:
Merge branch 'trunk' into fix/boost/page-cache-error-handling

pyronaur committed:
Add `useMutationNotice` hook

pyronaur committed:
Remove unused meta.tsx

pyronaur committed:
Show "setting up cache" notice earlier

pyronaur committed:
Also add idle and reset to MutationSubset

pyronaur committed:
Flip the useMutationNotice args for better devex

pyronaur committed:
Add js-packages/react-data-sync-client changelog entry.

pyronaur committed:
Clean up

pyronaur committed:
// todo

dilirity committed:
Update hook docblock comment

pyronaur committed:
Move auto-dismissal to manager.tsx

pyronaur committed:
Remove `hasNotice`

pyronaur committed:
removePageCacheNotice

pyronaur committed:
`isPending` > `isSuccess` - even if isSuccess exists, use `isPending`

So if a mutation happens over and over again, isPending is going to be re-triggered and notice is going to flip to isPending state again, even if `isSuccess` is stuck for some reason.

pyronaur committed:
Only auto-dismiss if a notice isn't pending

pyronaur committed:
`setNotice` is a dependency

Not sure if this is necessary, but looks like it doesn't hurt

pyronaur committed:
Clear the notice before turning the module off

pyronaur committed:
Set the timeout back to 5s

Commit:
Merge branch 'trunk' into fix/boost/page-cache-error-handling

thingalon commented on projects/plugins/boost/app/assets/src/js/features/page-cache/page-cache.tsx:
```diff
@@ -1,34 +1,290 @@
+	showErrorNotice: boolean;
+};
+
+const BypassPatterns = ( {
+	patterns,
+	setPatterns,
+	showErrorNotice = false,
+}: BypassPatternsProps ) => {
```
This file seems to contain multiple meaty components. Would it make sense to split it up?

Can be a separate PR.

thingalon commented on projects/plugins/boost/app/assets/src/js/features/ui/mutation-notice/mutation-notice.tsx:
```diff
@@ -1,34 +1,35 @@
- *		 savingMessage={ __( 'Saving…', 'jetpack-boost' ) }
- *		 errorMessage={ __(
- *		 'An error occurred while saving changes. Please, try again.',
- *		 'jetpack-boost'
- *		 ) }
- *		 successMessage={ __( 'Changes saved.', 'jetpack-boost' ) }
- *		 />
+ *	useMutationNotice( {
+ *		"unique-mutation-id",
+ *		mutationState,
+ *		{
+ *			savingMessage: 'Saving…',
+ *			errorMessage: 'An error occurred while saving changes. Please, try again.',
+ *			successMessage: 'Changes saved.',
+ *		}
+ *	} );
```
❤️

thingalon reviewed (approved):
This seems to work really well and this PR contains a bunch of improvements we very much want in. 

I have left about multiple complex components in one file - but that can be addressed in a separate PR. :)

thingalon commented:
Cherry-picked to 3.1.0 in a2795a9fb2.
